### DISCUSSION TITLE
Question about threadJoin and finalize in FlowChannelHeatStructureCouplerUserObject

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#27116] [Q&A General] [answered]
Hi all,

I'm confused about the function of `FlowChannelHeatStructureCouplerUserObject::threadJoin(const UserObject & uo)` and `FlowChannelHeatStructureCouplerUserObject::finalize()` in THM. 

#https://mooseframework.inl.gov/docs/doxygen/modules/FlowChannelHeatStructureCouplerUserObject_8C_source.html

From my point of view, `FlowChannelHeatStructureCouplerUserObject::execute()` can finish all we want and we do not need any information in threadJoin() and finalize(), why do we need the threadJoin() and finalize()? 

I just try to guess the reason:

The aim of FlowChannelHeatStructureCouplerUserObject is to construct a full getCachedQuantityMaps().

If we just have 1 thread, we do not need the threadJoin() and finalize() function.

Each MPI can include lots of threads. But if we have lots of threads and 1 MPI, for the threadJoin() function, we are not sure which threads calculate which elements. So we need to give the value of different threads to this main thread. After all the threadJoin() calculations, we could get a full map that can be used in Kernel, Material, and so on. In this case, finalize() is useless.

If we have lots of MPI, we do not know which MPI calculates which elements. In each MPI, we just get a part of the full map. So we need to gather all parts of the map together to construct a full map that can be used in Kernel, Material, and so on. So we need the finalize() function.

Is that right? Could anyone kindly tell me? Thanks!

**Comment by GiudGiud** [comment]
Hello

> If we just have 1 thread, we do not need the threadJoin()

yes this is correct. And 
```
If we just have 1 MPI process, we do not need the finalize()
```
this is often true but not always. Some user objects do guenuine work in the finalize routine. 

Everything you said makes sense to me. @joshuahansel  if you can double-check
This map is used for the purpose you mentioned and also for handling ghosting of DOFs across MPI processes. MPI processes need to know what DOF indices their immediate neighbor elements are on, the map helps with that.

Guillaume

**Reply by SomnusYu** [reply]
Thanks a lot!

**Comment by joshuahansel** [comment, marked as answer]
The UO needs to communicate the pieces of the map across threads and ranks because it executes only on the flow channel elements but is used by residual objects on both the flow channel and heat structure sides. So if for example the flow channel is on one thread/rank and the HS on the other, then you'd have a problem for the HS - the map entry wouldn't exist.

Everything you and Guillaume said is correct, except that `finalize()` may be necessary to implement even for 1 rank, as Guillaume pointed out (for example, a post-processor computing an average over all of the elements).

**Reply by SomnusYu** [reply]
I got it! Thanks a lot!
